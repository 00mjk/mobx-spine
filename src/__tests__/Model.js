import { Animal, Kind, Breed, Person, Location } from './fixtures/Animal';
import animalKindBreedData from './fixtures/animal-with-kind-breed.json';
import saveFailData from './fixtures/save-fail.json';
import axios from 'axios';
import { toJS } from 'mobx';
import MockAdapter from 'axios-mock-adapter';

test('Initialize model with valid data', () => {
    const animal = new Animal({
        id: 2,
        name: 'Monkey',
    });

    expect(animal.id).toBe(2);
    expect(animal.name).toBe('Monkey');
});

test('Initialize model with invalid data', () => {
    const animal = new Animal({
        nonExistentProperty: 'foo',
    });

    expect(animal.nonExistentProperty).toBeUndefined();
});

test('Initialize model without data', () => {
    const animal = new Animal(null);

    expect(animal.id).toBeNull();
});

test('URL should be correct without primary key', () => {
    const animal = new Animal();

    expect(animal.url).toBe('/api/animal/');
});

test('URL should be correct with primary key', () => {
    const animal = new Animal({ id: 2 });

    expect(animal.url).toBe('/api/animal/2/');
});

test('Relation should not be initialized by default', () => {
    const animal = new Animal();

    expect(animal.kind).toBeUndefined();
});

test('Initialize one-level relation', () => {
    const animal = new Animal(null, {
        relations: ['kind'],
    });

    expect(animal.kind).toBeInstanceOf(Kind);
});

test('isNew should be true for new model', () => {
    const animal = new Animal();

    expect(animal.isNew).toBe(true);
});

test('isNew should be false for existing model', () => {
    const animal = new Animal({ id: 2 });

    expect(animal.isNew).toBe(false);
});

test('Initialize two-level relation', () => {
    const animal = new Animal(null, {
        relations: ['kind.breed'],
    });

    expect(animal.kind).toBeInstanceOf(Kind);
    expect(animal.kind.breed).toBeInstanceOf(Breed);
});

test('Initialize three-level relation', () => {
    const animal = new Animal(null, {
        relations: ['kind.breed.location'],
    });

    expect(animal.kind).toBeInstanceOf(Kind);
    expect(animal.kind.breed).toBeInstanceOf(Breed);
    expect(animal.kind.breed.location).toBeInstanceOf(Location);
});

test('Initialize multiple relations', () => {
    const animal = new Animal(null, {
        relations: ['kind', 'owner'],
    });

    expect(animal.kind).toBeInstanceOf(Kind);
    expect(animal.owner).toBeInstanceOf(Person);
});

test('Initialize multiple nested relations', () => {
    const animal = new Animal(null, {
        relations: ['kind.breed', 'kind.location'],
    });

    expect(animal.kind.breed).toBeInstanceOf(Breed);
    expect(animal.kind.location).toBeInstanceOf(Location);
});

test('Attributes list', () => {
    const animal = new Animal();

    expect(animal._attributes).toEqual(['id', 'name']);
});

test('Parsing two-level relation', () => {
    const animal = new Animal(null, {
        relations: ['kind.breed'],
    });

    animal.fromBackend({
        data: animalKindBreedData.data,
        repos: animalKindBreedData.with,
        relMapping: animalKindBreedData.with_mapping,
    });

    expect(animal.id).toBe(1);
    expect(animal.name).toBe('Woofer');
    expect(animal.kind.id).toBe(4);
    expect(animal.kind.name).toBe('Good Dog');
    expect(animal.kind.breed.id).toBe(3);
    expect(animal.kind.breed.name).toBe('Good Pupper');
});

test('toBackend with basic properties', () => {
    const animal = new Animal({
        id: 3,
        name: 'Donkey',
    });

    const serialized = animal.toBackend();

    expect(serialized).toEqual({
        id: 3,
        name: 'Donkey',
    });
});

test('toBackend with relations', () => {
    const animal = new Animal({
        id: 4,
        name: 'Donkey',
    }, { relations: ['kind', 'owner'] });

    animal.kind.id = 8;

    const serialized = animal.toBackend();

    expect(serialized).toEqual({
        id: 4,
        name: 'Donkey',
        kind: 8,
        owner: null,
    });
});

test('clear with basic properties', () => {
    const animal = new Animal({
        id: 2,
        name: 'Monkey',
    });

    animal.clear();

    expect(animal.id).toBe(null);
    expect(animal.name).toBe('');
});

test('clear with relations', () => {
    const animal = new Animal({
        id: 5,
        name: 'Donkey kong',
    }, { relations: ['kind', 'owner'] });

    animal.kind.id = 8;

    animal.clear();

    expect(animal.kind.id).toBe(null);
});

describe('requests', () => {
    let mock;
    beforeEach(() => {
        mock = new MockAdapter(axios);
    });
    afterEach(() => {
        if (mock) {
            mock.restore();
            mock = null;
        }
    });

    test('fetch with basic properties', () => {
        const animal = new Animal({ id: 2 });
        mock.onAny().replyOnce((config) => {
            expect(config.url).toBe('/api/animal/2/');
            expect(config.method).toBe('get');
            expect(config.params).toEqual({ with: null });
            return [200, { data: { id: 2, name: 'Madagascar' } }];
        });

        return animal.fetch()
        .then(() => {
            expect(animal.id).toBe(2);
        });
    });

    test('fetch with relations', () => {
        const animal = new Animal({ id: 2 }, {
            relations: ['kind.breed'],
        });
        mock.onAny().replyOnce((config) => {
            expect(config.params).toEqual({
                with: 'kind.breed',
            });
            return [200, animalKindBreedData];
        });

        return animal.fetch()
        .then(() => {
            expect(animal.id).toBe(1);
            expect(animal.kind.id).toBe(4);
            expect(animal.kind.breed.id).toBe(3);
        });
    });

    test('save new with basic properties', () => {
        const animal = new Animal({ name: 'Doggo' });
        mock.onAny().replyOnce((config) => {
            expect(config.url).toBe('/api/animal/');
            expect(config.method).toBe('post');
            expect(config.data).toBe('{"name":"Doggo"}');
            return [201, { id: 10, name: 'Doggo' }];
        });

        return animal.save()
        .then(() => {
            expect(animal.id).toBe(10);
        });
    });

    test('save existing with basic properties', () => {
        const animal = new Animal({ id: 12, name: 'Burhan' });
        mock.onAny().replyOnce((config) => {
            expect(config.method).toBe('patch');
            return [200, { id: 12, name: 'Burhan' }];
        });

        return animal.save()
        .then(() => {
            expect(animal.id).toBe(12);
        });
    });

    test('save fail with basic properties', () => {
        const animal = new Animal({ name: 'Nope' });
        mock.onAny().replyOnce(400, saveFailData);

        return animal.save()
        .catch(() => {
            const valErrors = toJS(animal.backendValidationErrors);
            expect(valErrors).toEqual({
                name: ['This field cannot be blank.'],
                kind: ['This field cannot be null.'],
            });
        });
    });

    test('delete existing with basic properties', () => {
        const animal = new Animal({ id: 12, name: 'Burhan' });
        mock.onAny().replyOnce((config) => {
            expect(config.method).toBe('delete');
            expect(config.url).toBe('/api/animal/12/');
            return [204, null];
        });

        return animal.delete();
    });
});

