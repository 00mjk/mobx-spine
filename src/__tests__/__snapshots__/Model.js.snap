// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBackendAll with \`backendResourceName\` property model 1`] = `
Object {
  "data": Array [
    Object {
      "blaat": 2,
      "id": 1,
      "owners": Array [
        3,
      ],
    },
  ],
  "relations": Object {
    "kind": Array [
      Object {
        "id": 2,
      },
    ],
    "person": Array [
      Object {
        "id": 3,
      },
    ],
  },
}
`;

exports[`toBackendAll with deep nested relation 1`] = `
Object {
  "data": Array [
    Object {
      "id": -6,
      "kind": -7,
      "name": "",
    },
  ],
  "relations": Object {
    "breed": Array [
      Object {
        "id": -9,
        "location": -10,
        "name": "MyBreed",
      },
    ],
    "kind": Array [
      Object {
        "breed": -9,
        "id": -7,
        "location": -8,
        "name": "Aap",
      },
    ],
    "location": Array [
      Object {
        "id": -8,
        "name": "Apenheul",
      },
      Object {
        "id": -10,
        "name": "Amerika",
      },
    ],
  },
}
`;

exports[`toBackendAll with model relation 1`] = `
Object {
  "data": Array [
    Object {
      "id": 4,
      "kind": 5,
      "name": "",
      "owner": -6,
    },
  ],
  "relations": Object {
    "breed": Array [
      Object {
        "id": -5,
        "name": "",
      },
    ],
    "kind": Array [
      Object {
        "breed": -5,
        "id": 5,
        "name": "",
      },
    ],
    "owner": Array [
      Object {
        "id": -6,
        "name": "",
      },
    ],
  },
}
`;

exports[`toBackendAll with nested store relation 1`] = `
Object {
  "data": Array [
    Object {
      "id": -6,
      "name": "",
      "past_owners": Array [
        -7,
        -8,
      ],
    },
  ],
  "relations": Object {
    "past_owners": Array [
      Object {
        "id": -7,
        "name": "Henk",
        "town": -9,
      },
      Object {
        "id": -8,
        "name": "Krol",
        "town": -10,
      },
    ],
    "town": Array [
      Object {
        "id": -9,
        "name": "Eindhoven",
      },
      Object {
        "id": -10,
        "name": "Breda",
      },
    ],
  },
}
`;

exports[`toBackendAll with store relation 1`] = `
Object {
  "data": Array [
    Object {
      "id": -5,
      "name": "",
      "past_owners": Array [
        -6,
        -7,
        10,
      ],
    },
  ],
  "relations": Object {
    "past_owners": Array [
      Object {
        "id": -6,
        "name": "Bar",
      },
      Object {
        "id": -7,
        "name": "Foo",
      },
      Object {
        "id": 10,
        "name": "R",
      },
    ],
  },
}
`;
